src/java/org/apache/cassandra/index/sai/disk/v1/postings/MergePostingList.java

```diff
@@ -98,18 +113,40 @@ else if (next == lastRowId)
     @Override
     public long advance(long targetRowID) throws IOException
     {
-        temp.clear();
-
-        for (var peekable : postingLists)
+        // clean out obsolete child lists, and remember the smallest row seen in case
+        // we can use it for the fast path
+        long nextRowId = PostingList.END_OF_STREAM;
+        PostingList nextPostingList = null;
+        var it = postingLists.iterator();
+        while (it.hasNext())
         {
+            var peekable = it.next();
             peekable.advanceWithoutConsuming(targetRowID);
-            if (peekable.peek() != PostingList.END_OF_STREAM)
-                temp.add(peekable);
+            long peeked = peekable.peek();
+            // clean out obsolete child lists
+            if (peeked == PostingList.END_OF_STREAM)
+                it.remove();
+
+            if (lastRowId <= peeked && peeked < nextRowId)
+            {
+                nextRowId = peeked;
+                nextPostingList = peekable;
+            }
+        }
+
+        if (lastRowId == targetRowID) {
+            // we're asking for the next row, past the current row, which an arbitrary
+            // number of our child posting lists may be pointing to.  In this case we
+            // need to let the PQ do its work to figure out what the correct next row AFTER
+            // this one is.
+            return nextPosting();
         }
```

Comment:
Why don't we check this case first in the method? I think it's correct either way, but we'd avoid iterating through the whole `postingLists`.

Alternatively, if `nextRowId` is set just once in this method, we know that it is the only remaining child posting, so we could skip calling `nextPosting()` in that case and just end up with `return nextPostingList.nextPosting();`.

Reply:
Because we need to call advanceWithoutConsuming on everything before calling nextPosting.